jupyter_notebook/notebookapp.py

```diff
@@ -120,6 +128,36 @@ def load_handlers(name):
     mod = __import__(name, fromlist=['default_handlers'])
     return mod.default_handlers
 
+class ShimHandler(FileFindHandler):
+    def validate_absolute_path(self, *args, **kwargs):
+        return super(ShimHandler, self).validate_absolute_path(*args, **kwargs)
+
+    def get_content(self, *args, **kwargs):
+        self.log.warn('Deprecated widget Javascript path /static/widgets/js/*.js was used')
+        
+        # Get the content.  get_content will return either a bytestring
+        # or an iterator of bytestrings, depending on the file.
+        content = super(ShimHandler, self).get_content(*args, **kwargs)
+
+        # Check if something should be prepended.
+        if self.prepend is not None:
+            self.set_header("Content-Length", str(int(self._headers["Content-Length"]) + len(self.prepend)))
+            if isinstance(content, (bytes)): # Bytes
+                content = self.prepend + content
+            else: # Iterator
+                content = itertools.chain([self.prepend], content)
+        return content
+
+def make_shim_handler(log, prepend=None, encoding='utf-8'):
```

Comment:
You should only need a single ShimHandler. No need for this type-generating function.